File: sdks/node/src/plugins/kimi.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import {
  AGENT_KEY,
  buildKimiHook,
  kimiRuntimeContracts,
  removeManagedKimiHooks,
  renderKimiDocument,
  type KimiConfigDocument,
  type KimiHook,
} from './kimi-contract.js';
import {
  commitKimiInstallation,
  commitKimiUninstall,
  preflightKimiInstallation,
  prepareKimiInstallation,
  prepareKimiUninstall,
} from './kimi-installation.js';
import { kimiRuntimeFilesExist, readKimiDocuments } from './kimi-io.js';
import { SUPPORTED_AGENTS } from './registry.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

function installedHooks(
  document: KimiConfigDocument,
  guardPath: string,
  auditPath: string,
): KimiHook[] {
  return [
    ...removeManagedKimiHooks(document.hooks),
    buildKimiHook('PreToolUse', guardPath),
    buildKimiHook('PostToolUse', auditPath),
    buildKimiHook('PostToolUseFailure', auditPath),
  ];
}

export const kimiPlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const documents = await readKimiDocuments();
    await preflightKimiInstallation(config, documents);
  },

  async install(config: InstallConfig): Promise<void> {
    const documents = await readKimiDocuments();
    const paths = await preflightKimiInstallation(config, documents);
    const rendered = await Promise.all(documents.map((document) => renderKimiDocument(
      document,
      installedHooks(document, paths.guardPath, paths.auditPath),
    )));
    const prepared = await prepareKimiInstallation(config, rendered);
    await commitKimiInstallation(prepared);
    const runtimes = documents.map(({ contract }) => contract.runtimeName).join(' and ');
    console.log(
      `  ${runtimes}: global PreToolUse, PostToolUse, and PostToolUseFailure hooks installed.`,
    );
  },

  async uninstall(agentId?: string): Promise<void> {
    const documents = await readKimiDocuments();
    const rendered = await Promise.all(documents.map((document) => renderKimiDocument(
      document,
      removeManagedKimiHooks(document.hooks, agentId),
    )));
    await commitKimiUninstall(await prepareKimiUninstall(rendered));
  },

  async status(): Promise<PluginStatus> {
    const documents = await readKimiDocuments();
    const contracts = kimiRuntimeContracts(documents);
    const hookConfigured = contracts.length > 0;
    const hookScriptExists = hookConfigured && await kimiRuntimeFilesExist(contracts);
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: contracts.at(-1)?.configPath ?? documents[0].contract.configPath,
    };
  },
};

File: sdks/go/cmd/elydora/plugins/kimi.go
package plugins

import (
	"fmt"
)

// KimiPlugin manages Kimi Code and legacy kimi-cli global lifecycle hooks.
type KimiPlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Kimi commits its provider guard with the
// audit runtime and every detected user hook document.
func (p *KimiPlugin) ManagesGuardRuntime() bool {
	return true
}

// PreflightInstall validates every detected hook source and runtime identity.
func (p *KimiPlugin) PreflightInstall(config InstallConfig) error {
	documents, err := readAllKimiConfigs()
	if err != nil {
		return err
	}
	_, _, err = preflightKimiInstallation(config, documents)
	return err
}

func (p *KimiPlugin) Install(config InstallConfig) error {
	documents, err := readAllKimiConfigs()
	if err != nil {
		return err
	}
	paths, nodePath, err := preflightKimiInstallation(config, documents)
	if err != nil {
		return err
	}
	guardCommand, err := buildKimiCommand(nodePath, paths.guardPath)
	if err != nil {
		return err
	}
	auditCommand, err := buildKimiCommand(nodePath, paths.auditPath)
	if err != nil {
		return err
	}
	additions := make([]kimiHook, 0, 3)
	for _, item := range []struct{ event, command string }{
		{"PreToolUse", guardCommand},
		{"PostToolUse", auditCommand},
		{"PostToolUseFailure", auditCommand},
	} {
		hook, err := buildKimiHook(item.event, item.command)
		if err != nil {
			return err
		}
		additions = append(additions, hook)
	}
	rendered := make([]kimiRenderedDocument, 0, len(documents))
	for _, document := range documents {
		keep, err := keptKimiHookIndices(document.hooks, "")
		if err != nil {
			return err
		}
		change, err := renderKimiChange(document, keep, additions)
		if err != nil {
			return err
		}
		rendered = append(rendered, change)
	}
	changes, err := prepareKimiInstallationChanges(config, paths, rendered)
	if err != nil {
		return err
	}
	if err := writeKimiChanges(
		changes,
		"Install Kimi hooks",
		p.rename,
		paths.runtimeRoot,
		paths.agentDirectory,
	); err != nil {
		return err
	}
	fmt.Printf(
		"%s: global PreToolUse, PostToolUse, and PostToolUseFailure hooks installed.\n",
		kimiRuntimeNames(documents),
	)
	return nil
}

func (p *KimiPlugin) Uninstall(agentID string) error {
	documents, err := readAllKimiConfigs()
	if err != nil {
		return err
	}
	rendered := make([]kimiRenderedDocument, 0, len(documents))
	for _, document := range documents {
		keep, err := keptKimiHookIndices(document.hooks, agentID)
		if err != nil {
			return err
		}
		change, err := renderKimiChange(document, keep, nil)
		if err != nil {
			return err
		}
		rendered = append(rendered, change)
	}
	changes, err := prepareKimiUninstallChanges(rendered)
	if err != nil {
		return err
	}
	return writeKimiChanges(
		changes,
		"Uninstall Kimi hooks",
		p.rename,
		"",
		"",
	)
}

func (p *KimiPlugin) Status() (PluginStatus, error) {
	documents, err := readAllKimiConfigs()
	entry := SupportedAgents[kimiAgentKey]
	status := PluginStatus{
		AgentName: kimiAgentKey, DisplayName: entry.Name,
	}
	if err != nil {
		return status, err
	}
	status.ConfigPath = documents[0].contract.configPath
	contracts, err := kimiRuntimeContracts(documents)
	if err != nil {
		return status, err
	}
	if len(contracts) == 0 {
		return status, nil
	}
	status.ConfigPath = contracts[len(contracts)-1].configPath
	status.HookConfigured = true
	status.HookScriptExists, err = kimiRuntimeFilesExist(contracts)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}
